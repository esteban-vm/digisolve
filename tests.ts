import type { ReactElement } from 'react'
import { create, type ReactTestRendererJSON, type TestRendererOptions } from 'react-test-renderer'

const customCreate = (ui: ReactElement, options?: TestRendererOptions) => {
  return create(ui, options).toJSON() as ReactTestRendererJSON
}

export { customCreate as create }
export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'
