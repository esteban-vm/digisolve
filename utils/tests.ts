import type { ReactElement } from 'react'
import { create } from 'react-test-renderer'

const customCreate = (element: ReactElement) => create(element).toJSON()
export { customCreate as create }

export * from '@testing-library/react'
export * from 'jest-axe'
export { default as userEvent } from '@testing-library/user-event'
