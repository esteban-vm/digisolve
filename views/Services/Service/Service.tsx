import type { Component, PropsWithLink } from '@/types'
import { useRef } from 'react'
import { Waypoint } from 'react-waypoint'
import Link from 'next/link'
import { Icon, type IconName } from '@/components'
import { styled } from '@/styles'

export type ServiceProps = PropsWithLink<{ title: string; description: string; icon: IconName }>

const ServiceComponent: Component<ServiceProps> = ({ id, title, description, link = '#', icon, ...rest }) => {
  const serviceRef = useRef<HTMLElement>(null)

  const animateService = (action: 'add' | 'remove') => {
    return () => {
      serviceRef.current?.classList[action]('animate__zoomIn')
    }
  }

  return (
    <Waypoint onEnter={animateService('add')} onLeave={animateService('remove')}>
      <article aria-labelledby={id} ref={serviceRef} {...rest}>
        <Icon name={icon} size={3} className='icon' />
        <h4 id={id}>{title}</h4>
        <p>{description}</p>
        <Link href={link}>Read more</Link>
      </article>
    </Waypoint>
  )
}

ServiceComponent.defaultProps = {
  className: 'animate__animated animate__fast',
}

const Service = styled(ServiceComponent)`
  padding: 1% 1.5%;

  h4 {
    font-size: 150%;
    margin-bottom: 30px;
    font-weight: 400;
    text-transform: capitalize;
  }

  p {
    color: var(--color-grey);
    font-size: 90%;
    margin-bottom: 20px;
  }

  a {
    :link,
    :visited {
      color: var(--color-primary);
      text-transform: uppercase;
      text-decoration: none;
      outline: none;
      transition: opacity 200ms;
    }

    :hover {
      opacity: 0.7;
    }
  }

  .icon {
    margin-bottom: 30px;
    color: var(--color-secondary);
  }
`

export default Service
