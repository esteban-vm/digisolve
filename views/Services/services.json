[
  { "title": "branding", "icon": "tags" },
  { "title": "content", "icon": "pen-alt" },
  { "title": "marketing", "icon": "funnel-dollar" },
  { "title": "web design", "icon": "laptop-code" }
]
