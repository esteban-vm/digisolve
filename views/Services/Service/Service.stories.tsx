import type { Meta, StoryObj } from '@/stories'
import type { IconComponentProps } from '@/types'
import { Grid } from '@/components'
import { styled } from '@/styles'
import services from '../services.json'
import Service from './Service'

const Wrapper = styled.div`
  text-align: center;
  width: 50vw;

  article {
    background-color: var(--color-light);
  }
`

export default {
  argTypes: {
    icon: {
      control: false,
    },
  },
  component: Service,
  title: 'Views/Services/Services',
  decorators: [
    (Story, context) => {
      return (
        <Wrapper>
          <Grid.Row>
            <Grid.Col isOneThird>&nbsp;</Grid.Col>
            <Grid.Col isOneThird>
              <Story {...context} />
            </Grid.Col>
            <Grid.Col isOneThird>&nbsp;</Grid.Col>
          </Grid.Row>
        </Wrapper>
      )
    },
  ],
  parameters: {
    layout: 'centered',
    viewport: {
      disable: true,
    },
  },
} as Meta<typeof Service>

const Examples: StoryObj<typeof Service>[] = []
for (const service of services as IconComponentProps[]) Examples.push({ args: service })

export const Example1 = Examples[0]
export const Example2 = Examples[1]
export const Example3 = Examples[2]
export const Example4 = Examples[3]
