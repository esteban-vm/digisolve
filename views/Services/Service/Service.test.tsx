import { render, cleanup, screen, create, axe } from '@/tests'
import services from '../Services.data'
import Service from './Service'

describe('ðŸ§ª SERVICE:', () => {
  const [testService] = services

  describe('render tests:', () => {
    let container: HTMLElement

    beforeEach(() => {
      void ({ container } = render(<Service {...testService} />))
    })

    afterEach(cleanup)

    it('should be accessible', async () => {
      const results = await axe(container)
      expect(results).toHaveNoViolations()
    })

    describe('should be rendered:', () => {
      it('the title', () => {
        const title = screen.getByRole('heading', { name: testService.title, level: 4 })
        expect(title).toBeInTheDocument()
        expect(title).toBeVisible()
      })

      it('the description', () => {
        const desc = screen.getByText(testService.description)
        expect(desc).toBeInTheDocument()
        expect(desc).toBeVisible()
      })

      it('the link', () => {
        const link = screen.getByRole('link', { name: /^read more$/i })
        expect(link).toBeInTheDocument()
        expect(link).toBeVisible()
      })

      it('the icon', () => {
        const icon = screen.getByRole('img', { hidden: true })
        expect(icon).toBeInTheDocument()
        expect(icon).toBeVisible()
      })
    })
  })

  describe('style tests:', () => {
    let tree: ReturnType<typeof create>

    beforeEach(() => {
      tree = create(<Service {...testService} />)
    })

    it('should render with correct styles', () => {
      expect(tree).toMatchSnapshot()
    })

    it('should have correct styles when hovered', () => {
      expect(tree).toHaveStyleRule('opacity', '0.7', { target: 'a:hover' })
    })
  })
})
