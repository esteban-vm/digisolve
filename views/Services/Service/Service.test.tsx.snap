// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ðŸ§ª SERVICE: style tests: should render with correct styles 1`] = `
.emotion-0 {
  padding: 1% 1.5%;
}

.emotion-0 h4 {
  font-size: 150%;
  margin-bottom: 30px;
  font-weight: 400;
  text-transform: capitalize;
}

.emotion-0 p {
  color: var(--color-grey);
  font-size: 90%;
  margin-bottom: 20px;
}

.emotion-0 a:link,
.emotion-0 a:visited {
  color: var(--color-primary);
  text-transform: uppercase;
  -webkit-text-decoration: none;
  text-decoration: none;
  outline: none;
  -webkit-transition: opacity 200ms;
  transition: opacity 200ms;
}

.emotion-0 a:hover {
  opacity: 0.7;
}

.emotion-0 .icon {
  margin-bottom: 30px;
  color: var(--color-secondary);
}

<article
  aria-labelledby="99a36a03-7210-44c9-b031-f0e2acec54e5"
  className="animate__animated animate__fast emotion-0"
>
  <svg
    aria-hidden={true}
    className="icon"
    color="var(--color-primary)"
    fill="currentColor"
    height="3rem"
    role="img"
    viewBox="0 0 512 512"
    width="3rem"
  >
    <path
      d="M345 39.1l127.8 129.3c52.4 53 52.4 138.2 0 191.2l-112 113.3c-9.3 9.4-24.5 9.5-33.9.2s-9.5-24.5-.2-33.9l111.9-113.3c33.9-34.3 33.9-89.4 0-123.7L310.9 72.9c-9.3-9.4-9.2-24.6.2-33.9s24.6-9.2 33.9.2zM0 229.5V80c0-26.5 21.5-48 48-48h149.5c17 0 33.3 6.7 45.3 18.7l168 168c25 25 25 65.5 0 90.5L277.3 442.7c-25 25-65.5 25-90.5 0l-168-168C6.7 262.7 0 246.5 0 229.5zM144 144c0-17.7-14.3-32-32-32s-32 14.3-32 32 14.3 32 32 32 32-14.3 32-32z"
    />
  </svg>
  <h4
    id="99a36a03-7210-44c9-b031-f0e2acec54e5"
  >
    branding
  </h4>
  <p>
    Ipsam reiciendis eligendi est qui itaque.
  </p>
  <a
    href="#"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onTouchStart={[Function]}
  >
    Read more
  </a>
</article>
`;
